### ISSUE TITLE
A T9 lookalike

### ISSUE THREAD

**Issue by JassMan23** [ubports/keyboard-component#51]
### Description of the feature

**A "Single Tap" numeric keyboard for text entry.**

This would have some similarities with the Tegic T9 keyboard in that single taps are ambiguous
 until the entire word is typed. Unlike T9 there is no need for it to be predictive since prediction is
 already available in Maliit as a separate plugin. This keyboard would therefore correctly be termed
 reactive rather than predictive. Each keypress would result in a lookup from a word list based on all
 the keys pressed till that point. I would propose the use of 8 lookups to make wordlist management
 easier, each based on the first key pressed. eg 2dict.json, 3dict.json, .... 9dict.json

Other flavours of linux appear to have attempted to create what they call T9, but in fact all I have
 discovered so far are not T9 as described by Tegic in that they require mulitaps on each key to select
 individual letters. For this keyboard to avoid any suggestion it is based on T9, IMHO it should be called
 S9 ( _S_ingletap on keys 2 to 9).

This keyboard is therefore designed to:
- provide much larger keys than the standard OSK to prevent mistyping
- use less screen space because there are fewer rows of keys
- improve convergence because the proposed keypad has sufficient keys to allow for Ctrl, Alt etc.
 Indeed all keys from a 104 key extended keyboard can be emulated.

Illustrations are shown below to show proposed overlays. The main keyboard consists of 18 keys
 arranged as 3x6 (or 6x3 landscape). 10 numeric (with alpha on keys 2 to 9), 2 dedicated keys (Enter
 and Backspace/Delete), and 6 modifier/overlay_select.

Hopefuly most keys illustrated are self explanatory but some may require explanation:

- ">>>"/Alt is used to select the next alternative interpretation of a key sequence. A long keypress replicates Alt on standard PC.
- "-><-"/Ctrl is a deadspace key to create longer words by terminating the current lookup and starting an adjoined new word. Ctrl replicates Ctrl on a PC.
- "QikSym"/Esc opens an overlay of punctuation and symbols which returns to the main keypad after a single selection. Esc acts as on a standard PC, and should cancel dialogs without saving any changes.
- FKeys/Tab opens a FunctionKey overlay to allow replication of keys F1 to F12. Tab should act as on a standard PC.
- CursorPad/Compose provides similar functionality to longpress on Space on the OSK, but with additional nudge and action keys. Compose allows the creation of accented characters by switching to  multitap mode for selecting a character, then offering an accent selector.
- M9/1 swiches to multitap mode for entry on Pins and passwords. Longpress on 1 could allow a popup for punctuation selection.
- On the main keypad, a longpress on any numberkey should bring up a popup in a similar fashion to the standard OSK.

### Illustrations

Click to embiggen in a separate window (Ctrl **+** to enlarge)
![osk_portrait_overview](https://user-images.githubusercontent.com/42297153/48664142-a7731380-ea9a-11e8-882f-66237ad91f03.png)

And to show why we need a smaller keyboard:

![osk_landscape](https://user-images.githubusercontent.com/42297153/48663821-d5098e00-ea95-11e8-88a3-283bd0a0cc64.png)

extract from proposed 2dict.json
	"2253","able"
	"2253","bake"
	"2253","bald"
	"2253","bale"
	"2253","cake"
	"2253","calf"
	"2253243","caleche"
	"225326","calean"
	"22532742","Balearic"
Note the use of numbers as strings. This is because some libraries convert large integers to exponents. Antidisesttablishmentarianism would be corrupted and foreign languages such as German have even longer words.

**Comment by Fuseteam** [comment]
while not completely what your looking for flick covers some of the points you've mentioned

**Comment by JassMan23** [comment]
@Fuseteam - Any clues about "flick"? I have tried various search engines [including the built in search in OpenStore app] but I can find nothing other than flick() which is just an api as part of the  touch gestures.All other results seem to return flickr [the image site].

**Comment by Fuseteam** [comment]
@JassMan23 there's a test version of it in the QA group

I'll try to prepare a PR for it someday

**Comment by dobey** [comment]
@Fuseteam Perhaps what would be good here, is to describe what it is, perhaps with some links and/or video for reference, so one can determine if it's a suitable alternative for what is described in this feature request.

**Comment by Fuseteam** [comment]
I suppose @dobey is right
i don't have links or videos currently but i do have a picture

i can try to give a description:
flick is a "T9 look-alike" keyboard that works with "flicks"

It has the 12 keys present on the t9 keyboard and tapping it once inputs a character
this satisfies the suggestion for fewer bigger keys

unlike t9 the other characters are not inputted via multiple taps but instead by "flicking" the key left, up, down or right
"flicks" are short swipes starting on the key in one of those for directions.
As a result each key has the potential to input 5 characters, meaning you can input a total of 60 characters with just 12 keys.

the flick keyboard i have managed to derive from the japanese keyboard contains 20 keys (4x5)
12 keys that can input 60 characters
2 edge keys that can input an additional 8 characters
a casemodifier key that can handle changing case (before and after a letter has been inputted), input a space, input a carriage return and activate cursor mode
a backspace key to uh delete :p
2 cursor keys that can move the cursor backwards and forward aswell as activate cursor mode (yes that means this layout currently has 3 triggers to activate cursor mode)
and enter/commit key: the enter key turns into a commit key when in autocomplete mode to "commit" a string
and a language key to switch between layouts
as a result this keyboard, when based on qwerty, can input all the characters that can be inputed by a full sized physical qwerty keyboard without the need to switch layers (yes this includes punctuation)
![screenshot20190907_003632537](https://user-images.githubusercontent.com/10421851/64469635-b3a08e00-d10b-11e9-9c0a-3bc76ef60a18.png)

**Comment by JassMan23** [comment]
OK. After searching the web yet again and still finding no clues as to how to enable flick, I reread the above and decided that maybe that [keyboard-component/plugins/ja/](https://github.com/ubports/keyboard-component/tree/xenial/plugins/ja) might contain a clue. Sure enough, it was hidden really well. Having looked at the code and deciding I wouldn't be locked forever into a kanji keyboard that I would have no way of returning from, I took the plunge and installed Japanese as an available keyboard. It looks very little like the one posted above in that it runs a basic abc def ghi... pattern but it let me test the idea. Essentially, the flick process seems to work, although I found that sliding my fingers round all the time made my finger tips oversensitive. Also it looks like it would be impossible to achieve anything like the speeds you can achieve on a tegic T9 (not the multitap which some people call T9).   
A couple of bugs I noticed while playing - which may not be in the version pictured above:
* The top and bottom left hand keys, and third row right key have kanji annotations
* Flicking a little too high causes the keyboard to slide down and it is slightly difficult to slide it back up.
* The version of flick on my phone seems to have no shift key, although I can see there is one in the picture above.

**Comment by Fuseteam** [comment]
i'm sure what speeds you expecting but i am faster with the one i posted than with qwerty on mobile

to emphasize: the one i posted is a custom layout it is available as a zip in the QA group as well as pr 93
note that pr 93 depends on pr 86 
this due to depending on changes pr 86 makes to allow multiple layouts per language

not completely sure if the 'flicking too high' bug is actually due to flicking too high
i have noticed it happens sometimes when flicking up
the workaround appears to be to turn autocomplete/autocorrect off 
or restarting the keyboard with 'restart maliit-server'

**Comment by JassMan23** [comment]
I guess I was hoping I would be able to get to the speeds I got on my nokia 3330 and later on my 6820. I always thought the keying on old Nokia chocolate bar phones was so good that they should put it on laptops and save the cost of a full keyboard. I reckon it was as quick as desktop keyboard.    
Since I am not in the QA group do you have any idea when it will be available in a dev channel update?   
On another note, if it is in the en osk, how does one switch from qwerty to flick?   
Also can I suggest that it might be more logical if the flick key layout was the same as the annotion, ie. the horizontal bar of the popup has same order as the keytop. ie swipe up for number swipe sideways for qer or tyu and down for symbol.  I realise this means that the middle letter will be the default (non-flick) but it seems more logical to me.

**Comment by Fuseteam** [comment]
uh i have no idea when it will be available, as i've said it depends on another pr, it'll hopefully become available shortly after that pr is merged

if you want you can pm me on pm i can provide you with the files and steps neccesary if you wish to test it in the meantime
the custom layout is a bit more logical when it comes to the order
from the picture: left is q, up is e, right is r, tap is 1 and down is colon
it is that order for all keys

**Comment by Fuseteam** [comment]
also i'm pretty confident you can approach the speed you had with feature phones with this layout once you master this layout

**Comment by Fuseteam** [comment]
As google is rejecting the zip file lemme try to see if i can get it up here
[flick.zip](https://github.com/ubports/keyboard-component/files/3608150/flick.zip)
for completion here as the instructions as well

1. use Ubuntu Touch Tweak Tool to make rootfs r/w (you can reverse this when you're finished with step 3)
2. extract the zip (there should be a folder called en@flick)
3. open your terminal and use the command
```
sudo cp -a en@flick /usr/share/maliit/plugins/com/ubuntu/lib/
```
4. type `restart maliit-server`
5. choose en@flick in your keyboard setting

**Comment by Fuseteam** [comment]
@JassMan23 #93 is now installable via a few commands if you wish to try it
the steps are as follows:

1. run `sudo ubports-qa install keyboard-component 93`
2. select no if it asks to proceed
3. run `sudo apt install ubuntu-keyboard ubuntu-keyboard-english-flick && restart maliit-server`

note: step 2 and 3 is only neccesary if step 1 asks to "downgrade" or otherwise affect unity8
by selecting no in step 2 nothing is affected. the repo is now added thus enabling you to do step 3
if step 1 does not affect unity you don't need step 2 & 3, you do however need the `restart maliit-server` command
this appears to be a bug in the system it is not related to the changes in the layout
finally if it complains about storage space you can use the following command
`mkdir -p /home/phablet/apt && sudo mount --bind /home/phablet/apt /var/cache/apt`

**Comment by Fuseteam** [comment]
hey @JassMan23 the flick layout has evolved quite a bit since the last time, it is now installable from pr #128 with the following three commands, after you have remounted your rootfs (temporarily) as readwrite:
```
sudo ubports-qa install keyboard-component
sudo apt install ubuntu-keyboard-english-flick
restart maliit-server
```

**Comment by JassMan23** [comment]
Thanks @Fuseteam. I am currently working on a hardware project but will try this out as soon as I sober up after Xmas.

**Comment by peat-psuwit** [comment]
Hello,

Thank you for contributing to UBports. As part of project renaming and the effort to port Ubuntu Touch stack to Ubuntu 20.04, we're incrementally migrating repositories to GitLab.

Your issue is now migrated to:
https://gitlab.com/ubports/core/lomiri-keyboard/-/issues/51

Sorry for your inconvenience.
